New SeekableSource class for efficient file reading.

jkinkead opened the pull request:
On my laptop, this outperforms `scala.io.Source` reading character-wise by a factor of 3.4, and line-wise by a factor of 1.7.

If you want to see the slightly out-of-date version with a memory map option, you can check out [this commit](https://github.com/jkinkead/common/tree/2006bb1c2ee0851ef9c978a52347ff43a670cab5). That's a little slower for the line iterator, because it's not using the dirty tricks I gleaned from BufferedReader. It's still valid for comparing the memory map option with the standard option, though.

dirkgr commented on core/src/main/scala/org/allenai/common/SeekableSource.scala:
```diff
@@ -0,0 +1,256 @@
+  * @throws IllegalArgumentException if `bufferSize` is less than 3, or `codec` is not UTF-8 or
+  * ISO-8859-1
+  */
+class SeekableSource(inFile: FileChannel, bufferSize: Int = 8192)(implicit codec: Codec)
```
Couldn't you have a buffer much larger than this? Why not 8MB?

jkinkead replied:
I think I'm just living in the past . . . :P

This size is what `java.io.BufferedReader` uses. Otherwise, it's pretty arbitrary. I wasn't seeing huge boosts when I was playing around with sizes, but 8M seems like a totally sensible default.

dirkgr commented on core/src/main/scala/org/allenai/common/SeekableSource.scala:
```diff
@@ -0,0 +1,256 @@
+    val first = inBuffer.get()
+    val intVal = if ((first & 0x80) == 0) {
+      return first.toChar
+    } else if ((first & 0xe0) == 0xc0) {
```
This is conditional of the byte starting with `110`. Is that what you want? It's not what the comment says below.

jkinkead replied:
The comment is talking about secondary bytes (all bytes following the first size-indicating byte). I'll try to clarify (and address the TODO).

dirkgr commented on core/src/main/scala/org/allenai/common/SeekableSource.scala:
```diff
@@ -0,0 +1,256 @@
+      // character instead.
+      val second = inBuffer.get()
+      ((first & 0x1f) << 6) | (second & 0x3f)
+    } else if ((first & 0xf0) == 0xe0) {
```
And this starts with `1110`. Maybe say that somewhere?

jkinkead replied:
Done.

dirkgr commented on core/src/main/scala/org/allenai/common/SeekableSource.scala:
```diff
@@ -0,0 +1,256 @@
+        4
+      } else if ((first & 0xfe) == 0xfc) {
+        5
+      } else {
```
I read somewhere that UTF-8 was designed so you can pick up a stream in the middle. There is some way to figure out where the next character starts. If that's true, you shouldn't have to do this. On the other hand, we're in the realm of illegal UTF-8, so maybe that doesn't apply here anyways.

jkinkead replied:
Actually, yes, you can skip any bytes that start with the secondary-character sigil. I'll do that.

dirkgr commented on core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala:
```diff
@@ -0,0 +1,155 @@
+  implicit val codec = Codec.UTF8
+
+  /** Stores füü, in UTF-8. */
+  val foomlaut: Array[Byte] = Array('f'.toByte, 0xc3, 0xbc, 0xc3, 0xbc) map { _.toByte }
```
What about characters that span three bytes?

jkinkead replied:
Done.

dirkgr commented on core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala:
```diff
@@ -0,0 +1,155 @@
+    lines.next() should be("gaz")
+    lines.hasNext should be(false)
+  }
+}
```
What about a test for the malformed UTF-8 case?

jkinkead replied:
Done.

dirkgr commented:
I'm assuming you smoke-tested this on the Wikipedia corpus, to make sure it doesn't blow up?

jkinkead commented:
Yeah, my test numbers come from reading the wikipedia corpus - and verifying line counts with `scala.io.Source`.

jkinkead commented:
OK, comments addressed, and unit tests expanded! I also fixed a couple of bugs.

It turns out that Java *does* support all characters in UTF-16, so I updated the by-character reader to accommodate this. Did you want to take another look?

dirkgr commented:
UTF is messy business.

dirkgr commented:
When you have all the special cases handled, make sure it's still faster than what you're replacing :-)

jkinkead commented:
What's really strange is that the comments you made on the specific commits went to my gmail account . . . :)

I'll let you know when I've added tests for the other corner-cases (and fixed them if need be).

jkinkead commented:
Special-case tested (and handled).

dirkgr commented:
At this point, I have no further ideas about how to break this. LGTM!

jkinkead commented:
Sweet!
